lib/Dialect/ESI/runtime/cosim/lib/RpcServer.cpp

```diff
@@ -0,0 +1,355 @@
+                                   esi::cosim::VoidMessage *response) override;
+
+private:
+  int esiVersion;
+  std::vector<uint8_t> compressedManifest;
+  std::map<std::string, RpcServerReadPort *> readPorts;
+  std::map<std::string, RpcServerWritePort *> writePorts;
+
+  std::unique_ptr<Server> server;
+};
+using Impl = esi::cosim::RpcServer::Impl;
+
+//===----------------------------------------------------------------------===//
+// Read and write ports
+//
+// Implemented as simple queues which the RPC server writes to and reads from.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Implements a simple read queue. The RPC server will push messages into this
+/// as appropriate.
+class RpcServerReadPort : public ReadChannelPort {
+public:
+  RpcServerReadPort(Type *type) : ReadChannelPort(type) {}
+
+  bool read(MessageData &data) override {
+    std::optional<MessageData> msg = readQueue.pop();
+    if (!msg)
+      return false;
+    data = std::move(*msg);
+    return true;
+  }
+
+  utils::TSQueue<MessageData> readQueue;
+};
+
+/// Implements a simple write queue. The RPC server will pull messages from this
+/// as appropriate. Note that this could be more performant if a callback is
+/// used. This would have more complexity as when a client disconnects the
+/// outstanding messages will need somewhere to be held until the next client
+/// connects. For now, it's simpler to just have the server poll the queue.
+class RpcServerWritePort : public WriteChannelPort {
+public:
+  RpcServerWritePort(Type *type) : WriteChannelPort(type) {}
+  void write(const MessageData &data) override { writeQueue.push(data); }
+
+  utils::TSQueue<MessageData> writeQueue;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// RPC server implementations
+//===----------------------------------------------------------------------===//
+
+/// Start a server on the given port. -1 means to let the OS pick a port.
+Impl::Impl(int port) : esiVersion(-1) {
+  grpc::ServerBuilder builder;
+  std::string server_address("127.0.0.1:" + std::to_string(port));
+  // TODO: use secure credentials. Not so bad for now since we only accept
+  // connections on localhost.
+  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials(),
+                           &port);
+  builder.RegisterService(this);
+  server = builder.BuildAndStart();
+  if (!server)
+    throw std::runtime_error("Failed to start server on " + server_address);
+  writePort(port);
+  std::cout << "Server listening on 127.0.0.1:" << port << std::endl;
+}
+
+void Impl::stop() {
+  /// Shutdown the server and wait for it to finish.
+  server->Shutdown();
+  server->Wait();
+  server = nullptr;
+}
+
+Impl::~Impl() {
+  if (server)
+    stop();
+  for (auto &port : readPorts)
+    delete port.second;
+  for (auto &port : writePorts)
+    delete port.second;
```

Comment:
What's the argument for not using smart pointers for these?

Reply:
IIRC, I couldn't get them to compile at some point though now I can. Updated.